src/main/kotlin/uk/gov/dwp/dataworks/dks/encryptionmaterialsprovider/DKSEncryptionMaterialsProvider.kt

```diff
@@ -0,0 +1,129 @@
+
+        val httpClientProvider = SecureHttpClientProvider(identityStore!!, identityStorePassword!!, identityStoreAlias!!, identityKeyPassword!!,
+            trustStore!!, trustStorePassword!!)
+
+        return HttpKeyService(httpClientProvider, dataKeyServiceUrl!!)
+    }
+
+    fun initializeKeyservice(keyService: KeyService) {
+        this.keyService = keyService
+    }
+
+    override fun refresh() {}
+
+    override fun getEncryptionMaterials(): EncryptionMaterials {
+        throw UnsupportedOperationException("Secret Key pair is not initialised.")
+    }
+
+    override fun getEncryptionMaterials(materialsDescription: MutableMap<String, String>?): EncryptionMaterials {
+        val materialsDescriptionStr = materialsDescription?.entries?.map { "$it.key : ${it.value}" }?.joinToString("\n")
+        logger.info("Received materials description $materialsDescriptionStr")
+        val keyId = materialsDescription?.get(METADATA_KEYID)
+        val encryptedKey = materialsDescription?.get(METADATA_ENCRYPTED_KEY)
+        logger.info("Received keyId: '$keyId' and encryptedKey: '$encryptedKey' from materials description")
+        return if (null == keyId && null == encryptedKey) {
+            getMaterialForEncryption()
+        }
+        else {
+            getMaterialForDecryption(keyId, encryptedKey)
+        }
+    }
+
+    private fun getMaterialForEncryption(): EncryptionMaterials {
+        logger.info("Calling DKS to generate key")
+        val dataKeyResult = keyService.batchDataKey()
+        val decodeKey = Base64.getDecoder().decode(dataKeyResult.plaintextDataKey)
+        val secretKeySpec = SecretKeySpec(decodeKey, 0, decodeKey.size, ALGORITHM)
+        logger.info("DKS generated key successfully!")
+        val keyId = dataKeyResult.dataKeyEncryptionKeyId
+        val cipherKey = dataKeyResult.ciphertextDataKey
+        logger.info("Adding key id '$keyId' and cipher key '$cipherKey' to the S3 metadata")
+        return EncryptionMaterials(secretKeySpec)
+            .addDescription(METADATA_KEYID, keyId)
+            .addDescription(METADATA_ENCRYPTED_KEY, cipherKey)
+    }
+
+    private fun getMaterialForDecryption(keyId: String?, encryptedKey: String?): EncryptionMaterials {
+        logger.info("Calling DKS to decrypt key")
+        val decryptedKey = keyService.decryptKey(keyId!!, encryptedKey!!)
+        val decodeKey = Base64.getDecoder().decode(decryptedKey)
+        val secretKeySpec = SecretKeySpec(decodeKey, 0, decodeKey.size, ALGORITHM)
+        logger.info("DKS decrypted key successfully!")
+        return EncryptionMaterials(secretKeySpec)
+    }
+
+    private fun getDKSProperties(): Map<String, String> {
+        val prop = Properties()
+        try {
+            val inputStream = FileInputStream(DKS_PROPERTIES_PATH)
+            prop.load(inputStream)
+        }
+        catch (e: Exception) {
+            logger.error("Exception when loading DKS properties", e)
```

Comment:
Should we fail out here as the DKS won't be configured properly? I also think we should be checking that all the config items we want are present otherwise when we pass them to the `SSSecureHttpClientProvider` we'll have issues?